onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -35,21 +42,137 @@ static ncclDataType_t GetNcclDataType(onnxruntime::MLDataType type) {
   }
 }
 
+#ifndef USE_MPI
+namespace IPC {
+#define FLLOG LOGS_DEFAULT(VERBOSE)
+#define FLLOGERRNO LOGS_DEFAULT(WARNING) << "error:" << strerror(errno)
+
+int WriteOnRank0(ncclUniqueId* nccl_id, int word_size) {
+  int fd = socket(AF_INET,     /* network versus AF_LOCAL */
```

Comment:
The way you create socket is decades old.  Below is a new style:
```c++
/*
 * \param service_name port number(e.g. "80") or service name(e.g."http")
 */
int make_socket(const char *service_name) {
  int sfd;
  struct addrinfo hints;
  struct addrinfo *result, *rp;

  memset(&hints, 0, sizeof(struct addrinfo));
  hints.ai_family = AF_UNSPEC;     /* Allow IPv4 or IPv6 */
  hints.ai_socktype = SOCK_STREAM; /* TCP socket. use SOCK_DGRAM for UDP */
  hints.ai_flags = AI_PASSIVE;     /* For wildcard IP address */
  hints.ai_protocol = 0;           /* Any protocol */
  hints.ai_canonname = nullptr;
  hints.ai_addr = nullptr;
  hints.ai_next = nullptr;

  int s = getaddrinfo(nullptr, service_name, &hints, &result);
  if (s != 0) {
    fprintf(stderr, "getaddrinfo failed: %s\n", gai_strerror(s));
    return -1;
  }
  for (rp = result; rp != nullptr; rp = rp->ai_next) {
    sfd = socket(rp->ai_family, rp->ai_socktype, rp->ai_protocol);
    if (sfd == -1) continue;

    int on = 1;
    int rc = setsockopt(sfd, SOL_SOCKET, SO_REUSEADDR, (char *)&on, sizeof(on));
    if (rc < 0) {
      perror("setsockopt() failed");
      close(sfd);
      return -1;
    }

    int optval = 1;
    size_t optlen = sizeof(optval);
    if (setsockopt(sfd, SOL_SOCKET, SO_KEEPALIVE, &optval, optlen) < 0) {
      perror("setsockopt()");
      close(sfd);
      return -1;
    }
    optval = 5;
    if (setsockopt(sfd, IPPROTO_TCP, TCP_KEEPIDLE, &optval, optlen) < 0) {
      perror("setsockopt()");
      close(sfd);
      return -1;
    }
    optval = 5;
    if (setsockopt(sfd, IPPROTO_TCP, TCP_KEEPINTVL, &optval, optlen) < 0) {
      perror("setsockopt()");
      close(sfd);
      return -1;
    }

    fprintf(stdout, "SO_KEEPALIVE set on socket\n");
    rc = ioctl(sfd, FIONBIO, (char *)&on);
    if (rc < 0) {
      perror("ioctl() failed");
      close(sfd);
      return -1;
    }

    if (bind(sfd, rp->ai_addr, rp->ai_addrlen) == 0) break; /* Success */

    close(sfd);
  }

  if (rp == nullptr) { /* No address succeeded */
    fprintf(stderr, "Could not bind\n");
    return -1;
  }
  freeaddrinfo(result); /* No longer needed */
  return sfd;
}
```

Reply:
Sound good to use the NEW API, so we can support IPV6 later